### PR TITLE
feat: register MLS client

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Register an MLS client in `RegisterClientUseCase`

### Notes

- There some improvements which can be made when CoreCrypto API gets some fixes

- I'd like to refactor how we store keystore / proteus files in a separate PR.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterClientUseCase.kt [hunk 1]
```diff
@@ -64,9 +72,12 @@ class RegisterClientUseCaseImpl(
             RegisterClientResult.Failure.Generic(it)
         }, { registerClientParam ->
             clientRepository.registerClient(registerClientParam).flatMap { client ->
-                clientRepository.persistClientId(client.clientId).map {
-                    client
-                }
+                // TODO when https://github.com/wireapp/core-crypto/issues/11 is implemented we
+                // can remove registerMLSClient() and supply the MLS public key in registerClient().
+                mlsClientProvider.getMLSClient(client.clientId).flatMap { clientRepository.registerMLSClient(client.clientId, it.getPublicKey()) }
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterClientUseCase.kt:77]: Splitting into smaller chunks, maybe?

```suggestion
    override suspend operator fun invoke(
        password: String,
        capabilities: List<ClientCapability>?,
        preKeysToSend: Int
    ): RegisterClientResult = suspending {
        generateProteusPreKeys(preKeysToSend, password, capabilities).coFold({
            RegisterClientResult.Failure.Generic(it)
        }, { registerClientParam ->
            clientRepository.registerClient(registerClientParam).flatMap { client ->
                createMLSClient(client)
            }.fold({ failure ->
                if (failure is NetworkFailure.ServerMiscommunication && failure.kaliumException is KaliumException.InvalidRequestError)
                    when {
                        failure.kaliumException.isTooManyClients() -> RegisterClientResult.Failure.TooManyClients
                        failure.kaliumException.isMissingAuth() -> RegisterClientResult.Failure.InvalidCredentials
                        else -> RegisterClientResult.Failure.Generic(failure)
                    }
                else RegisterClientResult.Failure.Generic(failure)
            }, { client ->
                RegisterClientResult.Success(client)
            })
        })
    }

    private suspend fun createMLSClient(client: Client): Either<CoreFailure, Client> = suspending {
        // TODO when https://github.com/wireapp/core-crypto/issues/11 is implemented we
        // can remove registerMLSClient() and supply the MLS public key in registerClient().
        mlsClientProvider.getMLSClient(client.clientId)
            .flatMap { clientRepository.registerMLSClient(client.clientId, it.getPublicKey()) }
            .flatMap { keyPackageRepository.uploadNewKeyPackages(client.clientId) }
            .flatMap { clientRepository.persistClientId(client.clientId) }
            .map { client }
    }

    private suspend fun generateProteusPreKeys(
        preKeysToSend: Int,
        password: String,
        capabilities: List<ClientCapability>?
    ) = preKeyRepository.generateNewPreKeys(FIRST_KEY_ID, preKeysToSend).flatMap { preKeys ->
        preKeyRepository.generateNewLastKey().flatMap { lastKey ->
            Either.Right(
                RegisterClientParam(
                    password = password,
                    capabilities = capabilities,
                    preKeys = preKeys,
                    lastKey = lastKey
                )
            )
        }
    }
```

Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterClientUseCase.kt:77]: Side-note:

I wonder what kind of damage limitation we need to do for MLS if it ever fails during intermediate steps.

For example: Successfully register MLS Client, and create KeyPackages, but fails to persist upload key packages or create client id.

Would it keep working if we try again?

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Tiny suggestions and questions.

Awesome work! Looking forward to see this flying 🚀🚀

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterClientUseCase.kt (hunk 1 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterClientUseCase.kt:77]: Hmm the last thing we do is to persist the client ID so if that fails or any step before that, we would re-start the flow  from scratch by registering a new client with BE.

**Comment by typfel** [comment]
@vitorhugods ready to be reviewed again.

**Review by vitorhugods — APPROVED** [review]
💯

**Pull request merged**